locust/dispatch.py

```diff
@@ -0,0 +1,394 @@
+            # excess, this last iteration will stop those excess users.
+            self._dispatched_users.update(self._desired_users_assigned_to_workers())
+            self._effective_assigned_users.update(
+                {
+                    worker_node_id: {user_class: 0 for user_class in user_classes_count.keys()}
+                    for worker_node_id, user_classes_count in self._dispatched_users.items()
+                }
+            )
+
+        else:
+            ts_dispatch = time.perf_counter()
+
+            user_count_in_current_dispatch = 0
+
+            for i, user_class_to_add in enumerate(itertools.cycle(self._sorted_user_classes)):
+                # For large number of user classes and large number of workers, this assertion might fail.
+                # If this happens, you can remove it or increase the threshold. Right now, the assertion
+                # is there as a safeguard for situations that can't be easily tested (i.e. large scale distributed tests).
+                assert i < 5000, "Looks like dispatch is stuck in an infinite loop (iteration {})".format(i)
+
+                if self._all_users_have_been_dispatched:
+                    break
+
+                if all(
+                    self._dispatched_user_class_count(user_class) >= user_count
+                    for user_class, user_count in self._user_classes_count.items()
+                ):
+                    break
+
+                if self._dispatched_user_class_count(user_class_to_add) >= self._user_classes_count[user_class_to_add]:
+                    continue
+
+                if self._try_next_user_class_in_order_to_stay_balanced_during_ramp_up(user_class_to_add):
+                    continue
+
+                for j, worker_node_id in enumerate(itertools.cycle(sorted(self._effective_assigned_users.keys()))):
+                    assert j < int(
+                        2 * self._number_of_workers
+                    ), "Looks like dispatch is stuck in an infinite loop (iteration {})".format(j)
+                    if self._effective_assigned_users[worker_node_id][user_class_to_add] == 0:
+                        continue
+                    self._dispatched_users[worker_node_id][user_class_to_add] += 1
+                    self._effective_assigned_users[worker_node_id][user_class_to_add] -= 1
+                    user_count_in_current_dispatch += 1
+                    break
+
+                if user_count_in_current_dispatch == self._user_count_per_dispatch:
+                    break
+
+            self._assert_computation_duration_of_dispatch_is_reasonable(duration=time.perf_counter() - ts_dispatch)
+
+        return {
+            worker_node_id: dict(sorted(user_classes_count.items(), key=itemgetter(0)))
+            for worker_node_id, user_classes_count in sorted(self._dispatched_users.items(), key=itemgetter(0))
+        }
+
+    def _assert_computation_duration_of_dispatch_is_reasonable(self, duration: float) -> None:
```

Comment:
This feels flaky. Perhaps there is no way around it, but this sort of guard against unreasonable computation times should not be needed outside of testing.

Reply:
if this never happens in real life then maybe it is ok though.